File: app/src/main/java/blue/koenig/kingsfinances/model/database/PendingTable.java
package blue.koenig.kingsfinances.model.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteStatement;

import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.Item;
import com.koenig.commonModel.Operation;

import org.joda.time.DateTime;

import java.nio.ByteBuffer;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

import blue.koenig.kingsfinances.model.PendingOperation;
import blue.koenig.kingsfinances.model.PendingStatus;

/**
 * Created by Thomas on 25.11.2017.
 */

public class PendingTable extends Table<PendingOperation> {
    public static final String NAME = "pendings";
    public static final String COLUMN_DATE = "date";
    public static final String COLUMN_OPERATION = "operation";
    public static final String COLUMN_STATUS = "status";
    private List<String> columnNames;

    public PendingTable(SQLiteDatabase database, ReentrantLock lock) {
        super(database, lock);
        columnNames = new ArrayList<>(3);
        columnNames.add(COLUMN_DATE);
        columnNames.add(COLUMN_OPERATION);
        columnNames.add(COLUMN_STATUS);
    }

    @Override
    protected void setItem(ContentValues values, PendingOperation item) {
        values.put(COLUMN_DATE, dateTimeToValue(item.getDateTime()));
        values.put(COLUMN_OPERATION, byteableToValue(item.getOperation()));
        values.put(COLUMN_STATUS, item.getStatus().name());
    }

    @Override
    protected String getTableSpecificCreateStatement() {
        return "," + COLUMN_DATE + " LONG, " + COLUMN_OPERATION + " BLOB, " + COLUMN_STATUS + " TEXT";
    }

    @Override
    protected PendingOperation getItem(Cursor cursor) {
        Operation operation = getOperation(cursor, COLUMN_OPERATION);
        DateTime dateTime = getDateTime(cursor, COLUMN_DATE);
        PendingStatus status = PendingStatus.valueOf(getString(cursor, COLUMN_STATUS));
        return new PendingOperation(operation, status, dateTime);
    }

    private Operation getOperation(Cursor cursor, String columnOperation) {
        byte[] bytes = cursor.getBlob(cursor.getColumnIndex(columnOperation));
        return new Operation(ByteBuffer.wrap(bytes));
    }

    @Override
    public String getTableName() {
        return NAME;
    }

    @Override
    protected List<String> getColumnNames() {
        return columnNames;
    }

    @Override
    protected void bindItem(SQLiteStatement statement, Map<String, Integer> map, PendingOperation item) {
        statement.bindLong(map.get(COLUMN_DATE), dateTimeToValue(item.getDateTime()));
        statement.bindBlob(map.get(COLUMN_OPERATION), byteableToValue(item.getOperation()));
        statement.bindString(map.get(COLUMN_STATUS), item.getStatus().name());
    }

    public void updateStatus(PendingStatus status, String id) throws SQLException {
        List<String> columns = new ArrayList<>(1);
        columns.add(COLUMN_STATUS);
        update(id, columns, (statement, columnsMap) -> {
            statement.bindString(columnsMap.get(COLUMN_STATUS), status.name());
        });
    }

    public List<PendingOperation> getNonConfirmedOperations() {
        String query = COLUMN_STATUS + " != ?";
        String value = PendingStatus.CONFIRMED.name();
        return queryWithOneValue(query, value);
    }


}

File: app/src/main/java/blue/koenig/kingsfinances/model/database/PendingTable.kt
package blue.koenig.kingsfinances.model.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteStatement
import blue.koenig.kingsfinances.model.PendingOperation
import blue.koenig.kingsfinances.model.PendingStatus
import com.koenig.commonModel.Item
import com.koenig.commonModel.Operation
import com.koenig.commonModel.Operator
import java.nio.ByteBuffer
import java.sql.SQLException
import java.util.*
import java.util.concurrent.locks.ReentrantLock

/**
 * Created by Thomas on 25.11.2017.
 */

class PendingTable(database: SQLiteDatabase, lock: ReentrantLock) : ItemTable<PendingOperation<*>>(database, lock) {
    override val tableName: String
        get() = NAME
    override val tableSpecificCreateStatement: String
        get() = ",$COLUMN_DATE LONG, $COLUMN_OPERATION BLOB, $COLUMN_STATUS TEXT"
    override val columnNames: MutableList<String>
        get() = mutableListOf(COLUMN_DATE, COLUMN_OPERATION, COLUMN_STATUS)

    // oldest first
    val nonConfirmedOperations: List<PendingOperation<*>>
        get() {
            val query = COLUMN_STATUS + " != ? ORDER BY " + COLUMN_DATE
            val value = PendingStatus.CONFIRMED.name
            return queryWithOneValue(query, value)
        }

    override fun setItem(values: ContentValues, item: PendingOperation<*>) {
        values.put(COLUMN_DATE, dateTimeToValue(item.dateTime))
        values.put(COLUMN_OPERATION, byteableToValue(item.operation))
        values.put(COLUMN_STATUS, item.status.name)
    }


    override fun getItem(cursor: Cursor): PendingOperation<*> {
        val operation = getOperation(cursor, COLUMN_OPERATION)
        val dateTime = getDateTime(cursor, COLUMN_DATE)
        val status = PendingStatus.valueOf(getString(cursor, COLUMN_STATUS))
        return PendingOperation(operation, status, dateTime)
    }

    private fun getOperation(cursor: Cursor, columnOperation: String): Operation<*> {
        val bytes = cursor.getBlob(cursor.getColumnIndex(columnOperation))
        return Operation<Item>(ByteBuffer.wrap(bytes))
    }

    override fun bindItem(statement: SQLiteStatement, map: Map<String, Int>, item: PendingOperation<*>) {
        statement.bindLong(map[COLUMN_DATE]!!, dateTimeToValue(item.dateTime))
        statement.bindBlob(map[COLUMN_OPERATION]!!, byteableToValue(item.operation))
        statement.bindString(map[COLUMN_STATUS]!!, item.status.name)
    }

    @Throws(SQLException::class)
    fun updateStatus(status: PendingStatus, id: String) {
        val columns = ArrayList<String>(1)
        columns.add(COLUMN_STATUS)
        update(id, columns, { statement, columnsMap -> statement.bindString(columnsMap[COLUMN_STATUS]!!, status.name) })
    }


    fun addUpdate(item: Item, userId: String): Operation<Item> {
        val operation = Operation(Operator.UPDATE, item)
        addFrom(PendingOperation(operation), userId)
        return operation
    }

    companion object {
        val NAME = "pendings"
        val COLUMN_DATE = "day"
        val COLUMN_OPERATION = "operation"
        val COLUMN_STATUS = "status"
    }
}
